criu/files-reg.c

```diff
@@ -214,6 +219,72 @@ static int copy_file_to_chunks(int fd, struct cr_img *img, size_t file_size)
 
 		if (copy_chunk_from_file(fd, img_raw_fd(img), ce.off, ce.len))
 			return -1;
+		pr_perror("sendfile wrong: %ld+%ld", ce.off, ce.len);
+	}
+
+	return 0;
+}
+
+static int copy_file_to_chunks_fiemap(int fd, struct cr_img *img, size_t file_size)
+{
+	GhostChunkEntry ce = GHOST_CHUNK_ENTRY__INIT;
+	struct fiemap *fiemap;
+	int ret = 0;
+	long long extents_size;
+	long long i = 0;
+
+	fiemap = xzalloc(sizeof(struct fiemap));
+	if (!fiemap) {
+		pr_perror("Out of memory when allocating fiemap");
+		return -1;
+	}
+
+	fiemap->fm_length = ~0; /* Lazy */
+
+	/* Find out how many extents there are */
+	ret = ioctl(fd, FS_IOC_FIEMAP, fiemap);
+	if (ret < 0) {
+		xfree(fiemap);
+		pr_perror("fiemap ioctl() failed");
+		return ret;
```

Comment:
Oh, That's my miss, thanks for pointing that out for me.
Is that OK I just return -errno instead of -1?
```C
ret = ioctl(fd, FS_IOC_FIEMAP, fiemap_buf);
if (ret < 0) {
	ret = errno;
	goto out;
} else if (fiemap_buf->fm_mapped_extents == 0)
	goto out;
```

Reply:
Technically you can. Though I don't like returning custom exit codes from function which would not be used. We only care about errno == EOPNOTSUP, so please return -1 in all other cases, it would be much less confusing when we later read this code.

Reply:
Sure, I have changed it.